File: Session12/C++/BinaryTreeMaximumPathSum.cpp
https://leetcode.com/problems/binary-tree-maximum-path-sum/

class Solution {
    
    int findSum(TreeNode* root, int& res)
    {
        if(root==NULL)
        {
            return 0;
        }
        
        int leftSum = max(0, findSum(root->left, res));
        int rightSum = max(0, findSum(root->right, res));
        
        res = max(res, root->val + leftSum + rightSum);
        
        return root->val + max(leftSum, rightSum);
    }
    
public:
    int maxPathSum(TreeNode* root) {
        
        int res = INT_MIN;
        findSum(root, res);
        return res;
        
    }
};

File: Session12/Java/BinaryTreeMaximumPathSum.java
https://leetcode.com/problems/binary-tree-maximum-path-sum/

class Solution {
    
    int res = Integer.MIN_VALUE;
    
    int findSum(TreeNode root)
    {
        if(root==null)
        {
            return 0;
        }
        
        int leftSum = Math.max(0, findSum(root.left));
        int rightSum = Math.max(0, findSum(root.right));
        
        res = Math.max(res, root.val + leftSum + rightSum);
        
        return root.val + Math.max(leftSum, rightSum);
    }
    
    public int maxPathSum(TreeNode root) {
        
        findSum(root);
        return res;
        
    }
}